### PR TITLE
[NFC][ESIMD] Separate functions called from SYCL and ESIMD kernels

### PR DESCRIPTION
**Pull request by DenisBakhvalov**
This is just an NFC patch to collect functions reachable from
SYCL and ESIMD kernels. In later patch I will add reachability
analysis for every function, i.e. whether a function is reachable
from regular SYCL context, or ESIMD, or both.

Such reachabilty analysis will be used for semantic analysis and
later be stored in IR as metadata.

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -3222,8 +3242,11 @@ void Sema::MarkDevice(void) {
 
   for (Decl *D : syclDeviceDecls()) {
     if (auto SYCLKernel = dyn_cast<FunctionDecl>(D)) {
-      llvm::SmallPtrSet<FunctionDecl *, 10> VisitedSet;
-      Marker.CollectKernelSet(SYCLKernel, SYCLKernel, VisitedSet);
+
```
Comment (kbobrovs) [inline clang/lib/Sema/SemaSYCL.cpp:3245]: syclDeviceDecls includes SYCL_EXTERNAL functions, not only kernels, correct?
NIT:
I suggest to replace "Kernel" substring in the added code with something more accurate - TopLevelFunction, ExternalFunction, ExportedFunction,...

Comment (DenisBakhvalov) [reply clang/lib/Sema/SemaSYCL.cpp:3245]: > syclDeviceDecls includes SYCL_EXTERNAL functions, not only kernels, correct?

Yes, there is the corresponding code at the beginning of the `MarkDevice` function that handles that.

> NIT:
> I suggest to replace "Kernel" substring in the added code with something more accurate - TopLevelFunction, ExternalFunction, ExportedFunction,...

I agree. Here we collect a set of functions called from the SYCL/ESIMD kernel. So, maybe `collectSyclKernelFuncs` or `collectSyclKernelCallees`? And ESIMD counterparts, like `collectEsimdKernelCallees`.

Comment (kbobrovs) [reply clang/lib/Sema/SemaSYCL.cpp:3245]: My point is that the roots of the graph the collection works on include not only kernels, but also SYCL_EXTERNAL functions. So `Kernel` in the name is somewhat misleading.

Comment (DenisBakhvalov) [reply clang/lib/Sema/SemaSYCL.cpp:3245]: Ah, I see. The maybe `collectSyclCalleesSet`/`collectEsimdCalleesSet` ?

Comment (DenisBakhvalov) [reply clang/lib/Sema/SemaSYCL.cpp:3245]: The `SYCLKernel` variable name is itself wrong since it's not only kernels but also SYCL_EXTERNAL functions. :)

Comment (kbobrovs) [reply clang/lib/Sema/SemaSYCL.cpp:3245]: collectSyclCallGraphNodes/collectESIMDCallGraphNodes?

Comment (DenisBakhvalov) [reply clang/lib/Sema/SemaSYCL.cpp:3245]: Ok, I like your variant more.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -433,17 +433,36 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
...
+  // The set of functions called by a regular SYCL kernel.
+  llvm::SmallPtrSet<FunctionDecl *, 10> SyclCallGraphNodes;
+  // The set of functions called by an ESIMD kernel.
+  llvm::SmallPtrSet<FunctionDecl *, 10> EsimdCallGraphNodes;
   // The set of recursive functions identified while building the
   // kernel set, this is used for error diagnostics.
   llvm::SmallPtrSet<FunctionDecl *, 10> RecursiveSet;
+
+  // This is an interface function for collecting sets of
+  // functions called from an ESIMD kernel.
```
Comment (premanandrao) [inline clang/lib/Sema/SemaSYCL.cpp:445]:
```suggestion
  // functions in a SYCL call graph.
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 3]
```diff
@@ -433,17 +433,36 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
...
+  void collectSyclCallGraphNodes(FunctionDecl *RootNode) {
+    llvm::SmallPtrSet<FunctionDecl *, 10> VisitedSet;
+    collectCallGraphNodes(RootNode, RootNode, VisitedSet, SyclCallGraphNodes);
+  }
+
+  // This is an interface function for collecting sets of
+  // functions called from a regular SYCL kernel.
```
Comment (premanandrao) [inline clang/lib/Sema/SemaSYCL.cpp:452]:
```suggestion
  // functions in an ESIMD call graph.
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -433,17 +433,36 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
...
+  llvm::SmallPtrSet<FunctionDecl *, 10> SyclCallGraphNodes;
+  // The set of functions called by an ESIMD kernel.
+  llvm::SmallPtrSet<FunctionDecl *, 10> EsimdCallGraphNodes;
   // The set of recursive functions identified while building the
   // kernel set, this is used for error diagnostics.
   llvm::SmallPtrSet<FunctionDecl *, 10> RecursiveSet;
+
+  // This is an interface function for collecting sets of
+  // functions in a SYCL call graph.
+  void collectSyclCallGraphNodes(FunctionDecl *RootNode) {
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:446]: Are the 2 functions required? Can't we just do the SyclSimdAttr check in this function?

Comment (DenisBakhvalov) [reply clang/lib/Sema/SemaSYCL.cpp]: Thanks, fixed.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 5]
```diff
@@ -433,17 +433,33 @@ class MarkDeviceFunction : public RecursiveASTVisitor<MarkDeviceFunction> {
...
+private:
   // Determines whether the function FD is recursive.
   // CalleeNode is a function which is called either directly
   // or indirectly from FD.  If recursion is detected then create
   // diagnostic notes on each function as the callstack is unwound.
-  void CollectKernelSet(FunctionDecl *CalleeNode, FunctionDecl *FD,
-                        llvm::SmallPtrSet<FunctionDecl *, 10> VisitedSet) {
+  void collectCallGraphNodes(FunctionDecl *CalleeNode, FunctionDecl *FD,
+                             llvm::SmallPtrSet<FunctionDecl *, 10> VisitedSet,
+                             llvm::SmallPtrSet<FunctionDecl *, 10> &KernelSet) {
```
Comment (kbobrovs) [inline clang/lib/Sema/SemaSYCL.cpp:462]: NIT:
```suggestion
                             llvm::SmallPtrSet<FunctionDecl *, 10> &CollectedSet) {
```

**Review by kbobrovs — APPROVED** [review]
LGTM

**Pull request closed without merging**